common/state_machine.hpp

```diff
@@ -0,0 +1,231 @@
+		ctx_(ctx) {
+	}
+	StateMachineRunner(StateMachineRunner &) = delete;
+
+	void PostEvent(EventType event) override {
+		event_queue_.push(event);
+		PostToEventLoop();
+	}
+
+	void RunOne() {
+		if (event_queue_.empty()) {
+			return;
+		}
+
+		size_t size = event_queue_.size();
+		vector<State<ContextType, EventType> *> to_run;
+
+		for (size_t count = 0; count < size; count++) {
+			bool deferred = false;
+			auto event = event_queue_.front();
+			event_queue_.pop();
+			to_run.clear();
+
+			for (auto machine : machines_) {
+				typename StateMachine<ContextType, EventType>::TransitionCondition cond {
+					machine->current_state_, event};
+				if (machine->deferred_events_.find(event) != machine->deferred_events_.end()) {
+					deferred = true;
+				}
+
+				auto match = machine->transitions_.find(cond);
+				if (match == machine->transitions_.end()) {
+					// No match in this machine, continue.
+					continue;
+				}
+
+				auto &target = match->second;
+				to_run.push_back(target);
+				machine->current_state_ = target;
+			}
+
+			if (to_run.empty()) {
+				if (deferred) {
+					// Put back in the queue to try later. This won't be tried
+					// again during this run, due to only making `size`
+					// attempts in the for loop.
+					event_queue_.push(event);
+				} else {
+					log::Warning(
+						"State machine event " + to_string(static_cast<int>(event))
+						+ " was not handled by any state. This is a bug and could hang the state machine.");
+					assert(!to_run.empty());
+				}
+			} else {
+				for (auto &state : to_run) {
+					state->OnEnter(ctx_, *this);
+				}
+				// Since we ran something, there may be more events waiting to
+				// execute. OTOH, if we didn't, it means that all objects currently
+				// in the queue are deferred, and not actionable until at least one
+				// state machine reaches a different state.
+				if (!event_queue_.empty()) {
+					PostToEventLoop();
+				}
+				break;
+			}
+		}
+	}
+
+	// Continously run state machinery on event loop.
+	void AttachToEventLoop(events::EventLoop &event_loop) {
+		DetachFromEventLoop();
+
+		cancelled_ = make_shared<bool>(false);
+
+		event_loop_ = &event_loop;
+	}
+
+	void DetachFromEventLoop() {
+		if (cancelled_) {
+			*cancelled_ = true;
+			cancelled_.reset();
+		}
+		event_loop_ = nullptr;
+	}
+
+	void AddStateMachine(StateMachine<ContextType, EventType> &machine) {
+		machines_.push_back(&machine);
+	}
+
+private:
+	void PostToEventLoop() {
+		if (!event_loop_ || event_queue_.empty()) {
+			return;
+		}
+
+		auto cancelled = cancelled_;
+		event_loop_->Post([cancelled, this]() {
+			if (!*cancelled_) {
+				RunOne();
+			}
+		});
+	}
+
+	ContextType &ctx_;
+
+	shared_ptr<bool> cancelled_;
+	vector<StateMachine<ContextType, EventType> *> machines_;
+
+	queue<EventType> event_queue_;
+
+	// Would be nice with optional<EventLoop &> reference here, for better crash messages, but
+	// optional doesn't support references.
+	events::EventLoop *event_loop_;
```

Comment:
What about `weak_ref`?

Reply:
I take it you mean `weak_ptr`? It's not really what we need, because it's not the ownership we are tracking. What we are actually doing is:

```
bool event_loop_is_set_;
events::EventLoop *event_loop_;
```

but instead of a separate bool, we are taking `nullptr` to mean that the bool is false.

But I agree it could be improved with something similar. I'm changing it to take a shared pointer with a null deleter instead. That means we still don't own the object, but it will give us better messages if we ever use the null pointer by mistake.